### PR TITLE
Task/apps 3099 enroll device

### PR DESCRIPTION
**Pull request by anzioka**
PR:
- Handle device enrollment

### CODE REVIEW

File: chronicle/Models/Enrollment.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+//  Created by Alfonce Nzioka on 8/11/21.
+//
+
+import Foundation
+
+/// Describes user input values required to initiate device enrollment
+struct Enrollment {
+    var participantId: String = ""
+    var studyId: String = ""
+    var organizationId: String = ""
```
Comment (UnsungHero97) [inline chronicle/Models/Enrollment.swift]: Is there no `UUID` in swift?

File: chronicle/Models/IOSDevice.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    var name: String
+    var systemName: String
+    var localizedModel: String
+    
+    /// fully qualified java class name required when deserializing object
+    var className = "com.openlattice.chronicle.sources.IOSDevice"
+    
+    init(model: String, localizedModel: String, version: String, deviceId: String?, name: String, systemName: String) {
+        self.model = model
+        self.localizedModel = localizedModel
```
Comment (UnsungHero97) [inline chronicle/Models/IOSDevice.swift:25]: What is `localizedModel`? Do we need it?

File: chronicle/Models/IOSDevice.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+// information about the device
+// ref: https://developer.apple.com/documentation/uikit/uidevice
+struct IOSDevice :Codable {
+    var model: String
+    var version: String
+    var deviceId: String?
+    var name: String
+    var systemName: String
+    var localizedModel: String
```
Comment (UnsungHero97) [inline chronicle/Models/IOSDevice.swift]: Please sort.

File: chronicle/ViewModels/EnrollmentViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+//  EnrollmentViewModel.swift
+//  EnrollmentViewModel
+//
+//  Created by Alfonce Nzioka on 8/11/21.
+//
+
+import Foundation
+
+/// Grants view access to the Enrollment model.  Keeps track of input values and errors
+class EnrollmentViewModel: ObservableObject {
```
Comment (UnsungHero97) [inline chronicle/ViewModels/EnrollmentViewModel.swift:11]: What is a "ViewModel"? Consider renaming.

File: chronicle/chronicleApp.swift [hunk 5]
```diff
@@ -9,9 +9,19 @@ import SwiftUI
 
 @main
 struct chronicleApp: App {
```
Comment (UnsungHero97) [inline chronicle/chronicleApp.swift]: Why is `chronicleApp` (lowercase `c`) not `ChronicleApp`?

File: chronicle/Utilities/ApiUtils.swift [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        
+        var components = URLComponents()
+        
+        components.scheme = ChronicleApi.scheme
+        components.host = ChronicleApi.host
+        components.path = "\(ChronicleApi.v2Base)/\(enrollment.organizationId)/\(enrollment.studyId)/\(enrollment.participantId)/\(deviceId)\(ChronicleApi.enrollPath)"
+        
+        // debug: set components.scheme = 'http', components.host = [local server ip] , components.port = 8090
+        // expected path: /chronicle/v2 + ORGANIZATION_ID_PATH + STUDY_ID_PATH + PARTICIPANT_ID_PATH + DATASOURCE_ID_PATH + ENROLL_PATH
```
Comment (UnsungHero97) [inline chronicle/Utilities/ApiUtils.swift:42]: What's this?

File: chronicle/Utilities/ApiUtils.swift [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        
+        guard !deviceId.isEmpty else {
+            return nil
+        }
+        
+        var components = URLComponents()
+        
+        components.scheme = ChronicleApi.scheme
+        components.host = ChronicleApi.host
+        components.path = "\(ChronicleApi.v2Base)/\(enrollment.organizationId)/\(enrollment.studyId)/\(enrollment.participantId)/\(deviceId)\(ChronicleApi.enrollPath)"
```
Comment (UnsungHero97) [inline chronicle/Utilities/ApiUtils.swift]: Are you manually constructing the url path? Am I reading this correctly? If yes, I think this is fine for now but we should revisit in the future.

File: chronicle/chronicleApp.swift (hunk 5 above)
Comment (anzioka) [reply chronicle/chronicleApp.swift]: renamed

File: chronicle/Utilities/ApiUtils.swift (hunk 6 above)
Comment (anzioka) [reply chronicle/Utilities/ApiUtils.swift:42]: in dev mode, we want to hit the local chronicle server, thus we need to configure `components` accordingly.

File: chronicle/Models/IOSDevice.swift (hunk 2 above)
Comment (anzioka) [reply chronicle/Models/IOSDevice.swift:25]: https://developer.apple.com/documentation/uikit/uidevice/1620029-localizedmodel

used here: 
https://github.com/openlattice/chronicle-api/blob/main/src/main/java/com/openlattice/chronicle/sources/IOSDevice.kt

File: chronicle/ViewModels/EnrollmentViewModel.swift (hunk 4 above)
Comment (anzioka) [reply chronicle/ViewModels/EnrollmentViewModel.swift:11]: The class primarily handles state management, thus keeping views as stateless as possible. Would prefer to keep the name as it alludes to the MVVM architectural paradigm underlying the development of the app

File: chronicle/Utilities/ApiUtils.swift (hunk 7 above)
Comment (anzioka) [reply chronicle/Utilities/ApiUtils.swift]: Yes. Probably inevitable with swift

File: chronicle/Models/Enrollment.swift (hunk 1 above)
Comment (anzioka) [reply chronicle/Models/Enrollment.swift]: switched to UUID

**Review by UnsungHero97 — APPROVED** [review]

**Review by solomontang — APPROVED** [review]
would be great to see some tests

**Comment by anzioka** [comment]
> would be great to see some tests

https://github.com/openlattice/chronicle-ios/pull/3/files#diff-ce9d93131810771e893f6b689ec756aa74a3cc4d19eefeaa0457b17486ce3d0d

**Pull request merged**